Speed up AnimatedSwitcher.

gspencergoog opened the pull request:
This optimizes the AnimatedSwitcher so that it tags the right widget with its keyed subtree, and avoids rebuilding the transition unnecessarily.

This significantly improves the performance of Chips (which uses AnimatedSwitcher to swap out it's avatar and delete icon children).

Here's a graph of "before" and "after" this change for Chips (averaged over 10 seconds of frames for each number of chips: each chip has two animated switchers in it):
![chart 3](https://user-images.githubusercontent.com/8867023/39950441-f40df552-5535-11e8-80a0-ef88829d34f1.png)

gspencergoog committed:
Speed up AnimatedSwitcher.

Hixie commented:
Please add a test that verifies that if the AnimatedSwitcher has a stateful child, changing the child to a new staetful child does not cause the old child to lose its state as it's going away and the new child to reset its state once the old child is entirely gone.

Hixie commented:
(i suspect such a test would fail with this patch)

gspencergoog commented:
OK, I've reworked this to use KeyedSubtree.wrap with a serial number, eliminated the unnecessary recreation of the transition on every build, and implemented a cached list to keep the previous Widgets so that the list doesn't need to be rebuilt on each build.  I updated the perf graph, and it's slightly better than I had before (although not appreciably).  I added a test to make sure that stateful widgets weren't getting re-initialized with the transitions. PTAL.

gspencergoog commented on examples/flutter_gallery/lib/gallery/home.dart:
```diff
@@ -300,13 +300,6 @@ class _GalleryHomeState extends State<GalleryHome> with SingleTickerProviderStat
     super.dispose();
   }
 
-  static Widget _animatedSwitcherLayoutBuilder(List<Widget> children) {
```
I removed this because now that it has a center alignment (it was created originally with a different alignment), it is identical to the default builder, and isn't needed anymore.

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -48,9 +48,11 @@ typedef Widget AnimatedSwitcherTransitionBuilder(Widget child, Animation<double>
+/// The builder should return a widget which contains the given children, laid
+/// out as desired. It must not return null. The list of previous children is
+/// sorted from oldest to newest. The builder should be able to handle an empty
+/// list of previousChildren, or a null currentChild.
```
We should explicitly say that previousChildren does not contain currentChild.

Hixie replied:
We should also say if it's safe to mutate the previousChildren list, since a natural thing to do would be to add currentChild to that list if it's not null.

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -84,30 +85,33 @@ typedef Widget AnimatedSwitcherLayoutBuilder(List<Widget> children);
-///               key: new ValueKey<int>(_count),
-///               textScaleFactor: 3.0,
+///       child: new Directionality(
+///         textDirection: TextDirection.rtl,
```
Why is the text direction flipped here? Did you just want to reverse the column direction? If so, set an argument on the column.

Hixie replied:
(otherwise e.g. any text here will be the wrong direction)

Hixie replied:
Having it will make naive users cargo cult `Directionality`s into their code, which is worse because they don't get any error messages if they do that.

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -84,30 +85,33 @@ typedef Widget AnimatedSwitcherLayoutBuilder(List<Widget> children);
+///               },
+///               child: new Text(
+///                 '$_count',
+///                 // Must have this key to build a unique widget when _count changes.
```
it'll be a unique widget each time.
I think what you mean is "The key causes the AnimatedSwitcher to interpret this as a "new" child each time the count changes" or some such.

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -84,30 +85,33 @@ typedef Widget AnimatedSwitcherLayoutBuilder(List<Widget> children);
+///                 '$_count',
+///                 // Must have this key to build a unique widget when _count changes.
+///                 key: new ValueKey<int>(_count),
+///                 textScaleFactor: 3.0,
```
changing the text scale factor is a weird way of making the text bigger. Setting the textStyle is more idiomatic.

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -84,30 +85,33 @@ typedef Widget AnimatedSwitcherLayoutBuilder(List<Widget> children);
-///           ),
-///         ],
+///             new RaisedButton(
+///               child: new Text('Click!'),
```
accessibility folks will tell you to use a more descriptive label, like "Increment"

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -156,9 +160,10 @@ class AnimatedSwitcher extends StatefulWidget {
   /// the [child] in when the animation runs in the forward direction and out
-  /// when the animation runs in the reverse direction.
+  /// when the animation runs in the reverse direction. This is only called
+  /// when a new [child] is set (not for each build).
```
it should presumably be set any time the builder itself is changed, no?

ideally this should have the same semantics as the `Builder` widget.

Hixie replied:
> Are you saying that all the previous children should have their transitions updated when the transitionBuilder changes?

Yeah.

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -183,82 +188,97 @@ class AnimatedSwitcher extends StatefulWidget {
+  static Widget defaultLayoutBuilder(Widget currentChild, List<Widget> previousChildren) {
     return new Stack(
-      children: children,
+      children: currentChild == null ? previousChildren : previousChildren + <Widget>[currentChild],
```
allocating three lists each time looks expensive...

Hixie replied:
How about:

```dart
List<Widget> children = previousChildren;
if (currentChild != null)
  children = children.toList()..add(currentChild);
```

...then pass `children`.

gspencergoog replied:
OK, that works.  I inlined it though.

```dart
currentChild == null ? previousChildren : previousChildren.toList()..add(currentChild)
```

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -292,39 +310,24 @@ class _AnimatedSwitcherState extends State<AnimatedSwitcher> with TickerProvider
       _addEntry(animate: true);
-    } else {
-      if (_currentChild != null) {
-        _currentChild.widgetChild = widget.child;
```
if you remove this line, then surely you'll be making it impossible to update the child (in the case where it's not changing identity).

e.g. here's a test to add: create an AnimatedSwitcher, and set the child to `Text('1')`. Then, pump an isomorphic tree of widgets, but with the child set to `Text('2')`. Verify that the first time, the tree rendered `1`, and the second time it rendered `2` with no animation (and no `1`).

gspencergoog committed:
Review changes

gspencergoog committed:
Review changes

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -1342,18 +1342,22 @@ class _RawChipState extends State<RawChip> with TickerProviderStateMixin<RawChip
       ),
-      widget.deleteButtonTooltipMessage ?? MaterialLocalizations.of(context)?.deleteButtonTooltip,
-      widget.onDeleted,
+      hashValues(iconThemeData, widget.deleteIcon),
```
hashValues isn't unique. This will cause flakes.

Hixie replied:
As far as I can tell you don't really need to have a key here. What matters is whether there is a button or not; not what colour it is (e.g. imagine how this will work when the theme animates).

Hixie replied:
maybe key it based on a new kind of key, say `WidgetIdentityKey`, that takes a _Widget and delegates to canUpdate to compare instances...

Hixie replied:
(and just pass deleteIcon)

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -183,82 +195,105 @@ class AnimatedSwitcher extends StatefulWidget {
-  final Set<_AnimatedSwitcherChildEntry> _children = new Set<_AnimatedSwitcherChildEntry>();
+  final Set<_AnimatedSwitcherChildEntry> _previousChildren = new Set<_AnimatedSwitcherChildEntry>();
   _AnimatedSwitcherChildEntry _currentChild;
+  List<Widget> _previousChildWidgetCache = new List<Widget>.unmodifiable(<Widget>[]);
```
`const <Widget>[]`

Hixie replied:
(...is unmodifiable)

gspencergoog replied:
Oh, cool, so `new List<Widget>.unmodifiable(<Widget>[])` is equivalent to `const <Widget>[]`?  Done.

Hixie replied:
for all intents and purposes, yeah.

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -292,39 +325,37 @@ class _AnimatedSwitcherState extends State<AnimatedSwitcher> with TickerProvider
     } else {
+      // Make sure we update the child widget in _currentChild even if we're not
+      // going to start a new animation, but keep the key from the previous
+      // transition, since otherwise there is a huge performance penalty.
```
s/since otherwise there is a huge performance penalty//. If you don't keep the key, then you're not going to update the child, you're going to replace it.

gspencergoog replied:
Done

gspencergoog committed:
Making switcher update transitions when transition builder is set.

gspencergoog commented:
OK, made it so that transitions are updated when the transitionBuilder changes. PTAL

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -968,10 +954,9 @@ class ActionChip extends StatelessWidget implements ChipAttributes, TappableChip
-        ),
+            onPressed != null,
+            'Rather than disabling an ActionChip by setting onPressed to null, '
+            'remove it from the interface entirely.',),
```
style-o

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -1342,13 +1316,14 @@ class _RawChipState extends State<RawChip> with TickerProviderStateMixin<RawChip
     return _wrapWithTooltip(
       new InkResponse(
         onTap: widget.isEnabled ? widget.onDeleted : null,
         child: new IconTheme(
```
we should always put child last, makes it clearer what the later arguments (data, etc) relate to.

gspencergoog replied:
I agree... but isn't that how this is already?

Hixie replied:
oh my i misread this even more than i thought.

I meant for _wrapWithTooltip.

gspencergoog replied:
Ahh, I see what you mean now.  Fixed.

Hixie commented on packages/flutter/lib/src/material/chip.dart:
```diff
@@ -1628,15 +1603,15 @@ class _ChipRenderTheme {
+        typedOther.labelPadding == labelPadding &&
+        typedOther.showAvatar == showAvatar &&
+        typedOther.showCheckmark == showCheckmark &&
+        typedOther.canTapBody == canTapBody;
```
old style was more readable

gspencergoog replied:
ACK.  Sorry, I hit the 'dartfmt' key, and forgot to go back and revert the spurious changes that don't match Flutter style.  Most of the changes in this file were due to that: I've reverted them.

Hixie replied:
fwiw i'm agnostic to the ones i didn't comment on (most of them)

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -48,9 +48,14 @@ typedef Widget AnimatedSwitcherTransitionBuilder(Widget child, Animation<double>
+/// The previousChildren list is an unmodifiable list, sorted with the oldest
+/// at the beginning and the newest at the end. It does not include the
+/// currentChild.
+typedef Widget AnimatedSwitcherLayoutBuilder(Widget currentChild, List<Widget> previousChildren);
```
argument names in these paragraphs should be marked with backticks.

gspencergoog replied:
Done.

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -63,51 +68,55 @@ typedef Widget AnimatedSwitcherLayoutBuilder(List<Widget> children);
 /// class ClickCounter extends StatefulWidget {
 ///   const ClickCounter({Key key}) : super(key: key);
-///
+/// 
```
?

gspencergoog replied:
I was being space-y.

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -63,51 +68,55 @@ typedef Widget AnimatedSwitcherLayoutBuilder(List<Widget> children);
 /// class _ClickCounterState extends State<ClickCounter> {
 ///   int _count = 0;
-///
+/// 
```
?

gspencergoog replied:
Fixed

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -63,51 +68,55 @@ typedef Widget AnimatedSwitcherLayoutBuilder(List<Widget> children);
-///               // Must have this key to build a unique widget when _count changes.
-///               key: new ValueKey<int>(_count),
-///               textScaleFactor: 3.0,
+///       child: new Directionality(
```
see earlier discussion

gspencergoog replied:
I changed the example to just use a MaterialApp instead of Directionality.

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -156,9 +165,11 @@ class AnimatedSwitcher extends StatefulWidget {
-  /// when the animation runs in the reverse direction.
+  /// when the animation runs in the reverse direction. This is only called
+  /// when a new [child] is set (not for each build). The function must not
+  /// return null.
```
these docs are now out of date

gspencergoog replied:
Fixed.

Hixie commented on packages/flutter/lib/src/widgets/animated_switcher.dart:
```diff
@@ -215,50 +228,75 @@ class AnimatedSwitcher extends StatefulWidget {
+  }
+
+  void _rebuildChildWidgetCache() {
+    _previousChildWidgetCache = new List<Widget>.unmodifiable(
```
we call this in various places currently, with the effect that one frame could in theory invoke it multiple times.
I recommend instead just marking the cache as dirty (by setting it to null) and building it on demand (if null) in the build method (where you use it).

gspencergoog replied:
OK, good idea. Done.

Hixie commented:
How are the numbers now that I've nit picked this to death? Do you still get the benefit that this was intended to get?

gspencergoog commented:
Yes, the numbers haven't changed: It still looks like the graph at the top.

gspencergoog commented:
(I've been checking that after every change)

Hixie commented:
![LGTM](http://junkyard.damowmow.com/577)
